src/Microsoft.AspNetCore.JsonPatch/Internal/ObjectVisitor.cs

```diff
@@ -49,21 +51,20 @@ public bool TryVisit(ref object target, out IAdapter adapter, out string errorMe
 
         private IAdapter SelectAdapater(object targetObject)
         {
-            if (targetObject is ExpandoObject)
-            {
-                return new ExpandoObjectAdapter();
-            }
-            else if (targetObject is IDynamicMetaObjectProvider)
+            var jsonContract = _contractResolver.ResolveContract(targetObject.GetType());
+
+            if (targetObject is IList)
             {
-                return new DynamicObjectAdapter();
+                return new ListAdapter();
             }
-            else if (targetObject is IDictionary)
+            else if (jsonContract is JsonDictionaryContract jsonDictionaryContract)
             {
-                return new DictionaryAdapter();
+                var type = typeof(DictionaryAdapter<,>).MakeGenericType(jsonDictionaryContract.DictionaryKeyType, jsonDictionaryContract.DictionaryValueType);
+                return (IAdapter)Activator.CreateInstance(type);
             }
```

Comment:
We may still need to handle the case where an object is an `IDictionary` but not an `IDictionary<>`

Reply:
We've decided not to do anything about this.